src/Workspaces/Remote/Core/Services/RemoteWorkspace.cs

```diff
@@ -0,0 +1,160 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using Microsoft.CodeAnalysis.Host;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    /// <summary>
+    /// primary workspace for remote host. no one except solution service can update this workspace
+    /// </summary>
+    internal class RemoteWorkspace : Workspace
+    {
+        public const string WorkspaceKind_RemoteWorkspace = "RemoteWorkspace";
+
+        // REVIEW: I am using semaphoreSlim since workspace is using it, but not sure why it uses
+        //         semaphore rather than just object since workspace is not using anything specific
+        //         to semaphore
+        private readonly SemaphoreSlim _serializationLock = new SemaphoreSlim(initialCount: 1);
```

Comment:
Doesn't the workspace use async blocking on the SemaphoreSlim?  if not, we can change them to be normal objects.